contracts/tips/src/contract.rs

```diff
@@ -0,0 +1,2252 @@
+        service_fee.check_fees(&info.funds, &tip_amount)?;
+    }
+
+    // Resolve the receiver and the optional post id
+    let (post_id, receiver) = match target {
+        Target::ContentTarget { post_id } => {
+            let post = PostsQuerier::new(deps.querier.deref())
+                .query_post(config.subspace_id, post_id.u64())
+                .map_err(|_| ContractError::PostNotFound { id: post_id.u64() })?
+                .post;
+            (post.id.u64(), post.author)
+        }
+        Target::UserTarget { receiver } => {
+            // Set the post id to 0 (invalid id) to signal that this tip is referencing an user
+            (0_u64, deps.api.addr_validate(&receiver)?)
+        }
+    };
+
+    if info.sender == receiver {
+        return Err(ContractError::SenderEqReceiver {});
+    }
+
+    if config.tips_history_size > 0 {
+        // Generates the tip key
+        let tip_key = BLOCK_INDEX.update::<_, ContractError>(deps.storage, |(block, index)| {
+            if block == env.block.height {
+                Ok((
+                    block,
+                    index
+                        .checked_add(1)
+                        .ok_or(ContractError::BlockIndexOverflow {})?,
+                ))
+            } else {
+                Ok((env.block.height, 0))
+            }
+        })?;
+
+        // Build the tip that will be save inside the contract state
+        let tip = StateTip {
+            sender: info.sender.clone(),
+            receiver: receiver.clone(),
+            amount: tip_amount.clone(),
+            post_id,
+            ref_counter: if post_id > 0 {
+                3 // Sender + Receiver + Post histories
+            } else {
+                2 // Sender + Receiver histories
+            },
+        };
+
+        // Save the tip
+        TIPS.save(deps.storage, tip_key.clone(), &tip)?;
+        // Update sender tips history
+        add_tip_to_subject_history(
+            deps.storage,
+            &SENT_TIPS_HISTORY,
+            tip.sender,
+            tip_key.clone(),
+            config.tips_history_size,
+        )?;
+        // Update receiver tips history
+        add_tip_to_subject_history(
+            deps.storage,
+            &RECEIVED_TIPS_HISTORY,
+            tip.receiver,
+            tip_key.clone(),
+            config.tips_history_size,
+        )?;
+
+        // Tip referencing a post, let's update the post's tip history
+        if tip.post_id > 0 {
+            // Update receiver tips history
+            add_tip_to_subject_history(
+                deps.storage,
+                &POST_TIPS_HISTORY,
+                tip.post_id,
+                tip_key,
+                config.tips_history_size,
+            )?;
+        }
+    }
+
+    Ok(Response::new()
+        .add_attribute(ATTRIBUTE_ACTION, ACTION_SEND_TIP)
+        .add_attribute(ATTRIBUTE_SENDER, info.sender)
+        .add_attribute(ATTRIBUTE_RECEIVER, receiver.as_str())
```

Comment:
It would be useful to have `ATTRIBUTE_TIP_AMOUNT` with value like `100udsm,100uatom` so that other contract can be easier to integrate with this method.
What do you think?

Reply:
I will add it!

Reply:
@dadamu I have also added an `ATTRIBUTE_TIP_POST_ID` for when the tip is for a post.